### ISSUE TITLE
use argparse for the cli; make it more flexible

### ISSUE THREAD

**Issue by zenhack** [CCI-MOC/hil#431]
Per my comments in #425, the cli code is somewhat brittle; there's a standard library module `argparse` that we should be using instead of the stuff that I hacked together a year and a half ago.

**Comment by SahilTikale** [comment]
Added support for same, submitted the pull request today  #425

**Comment by zenhack** [comment]
Just a thought: I've been playing with [click][1] lately, and it's pretty nice -- more pleasant to use than argparse IMO. Flask also has some integration with it, though I'm not sure we actually want to use that. Perhaps it's worth considering using that instead of argparse.

[1]: http://click.pocoo.org

**Comment by zenhack** [comment]
In particular, it's handling of subcommands is much nicer.

**Comment by henn** [comment]
FYI - once we have argparse (or something functionally equivalent), we can do stuff like accept the username/password on the command line or have it [automatically read from a file](https://docs.python.org/2/library/argparse.html#fromfile-prefix-chars).

**Comment by zenhack** [comment]
Re: dealing with the CLI, I think one thing we'd talked about is waiting until we had a client library in place (see #617), and building the new CLI on top of that. @henn, what are your thoughts?

**Comment by shwsun** [comment]
My $0.02 is to merge the cli library we are working on first, then talk about this one. We also have to figure out whether or not implementing cli with click is our top priority.

**Comment by matsuuran** [comment]
I wanted to do it because @henn asked me to do #631 and ideally it would just use an optional param.  I'll do whatever though

**Comment by zenhack** [comment]
Quoting (Jethro) Shuwen Sun (2016-09-07 16:48:22)

>    My $0.02 is to merge the cli library we are working on first, then talk
>    about this one. We also have to figure out whether or not implementing
>    cli with click is our top priority.

I think the questions to be asking are:

1. How high priority is 631?
2. Are there any other high-priority issues that depend on this?
3. How close are we to actually being able to merge the client library?
   My memory is "not very."
4. How much work will we save by doing the client library first, rather
   than having to glue the CLI to it later?

**Comment by shwsun** [comment]
Of course, another option is to let @matsuuran start with the new one first, but making sure that it will not break the cli library we will merge later.

@SahilTikale is preempted by the NSDI / BMI work so it might be a little hard for him to finish the work quick. I will bring this up in tmr's hil meeting.

// cc @henn

**Comment by SahilTikale** [comment]
There was hick up with client library ( mostly difference of opinion) which has been settled.
1. I just need to fix the merge issues, 
2. Get keystone support copying stuff from #626 
3. responds to changes suggested by the reviewers.

I will insist on waiting to work on CLI until client library is done.

**Comment by zenhack** [comment]
Per discussion in #337, we most likely want to have a `--config` option that can specify the config file location. I'm mentioning it here so we can close that issue without losing the information.

**Comment by naved001** [comment]
@vsemp

**Comment by Izhmash** [comment]
I'm going to start working on this -- although I'm part-time right now I'll be able to chip away at it each workday.

**Comment by Izhmash** [comment]
Unassigning myself from this for now, other tasks are taking up most of my part-time hours.

**Comment by zenhack** [comment]
Sounds good.

So it's recorded here: we've decided to fold #515 into this issue; see the discussion there.

**Comment by naved001** [comment]
I started working on the CLI using argparse. Here's the [diff](https://github.com/CCI-MOC/hil/compare/master...naved001:introduce-argparse?expand=1) @zenhack if you have time could you look at the list methods in the new cli (there are some comments in those methods)? I am trying to figure out the right way to do it before adding more calls. Just any general feedback would be cool, Thanks!

**Comment by zenhack** [comment]
General feedback:

* I agree we should move `serve` to `hil-admin`
* We should probably take the opportunity to move the hil script over to the entry_points mechanism, per the TODO in setup.py.
* A class seems excessive.
* I like the idea of having a top-level `list` with its own subcommands per-object.
* I think we should do the same with `show`
* I would recommend the use of subcommands instead:

https://docs.python.org/2/library/argparse.html#sub-commands

Otherwise looks sensible.

**Comment by naved001** [comment]
> A class seems excessive.

When I have a class, I can check for invalid commands using `hasattr` and dispatch using `getattr`. 
Without it, I have the following options:

1. Use `globals()` to dispatch, but for checking the validity it's not a good idea since it includes everything in the module. 

2. Use a decorator to build to dictionary of valid commands (just like the current cli) and dispatch calls using `globals()`.

Any other options?

**Comment by zenhack** [comment]
I would use the subcommands mechanism provided by argparse, rather than trying to do it in an ad-hoc way.

re (1): you have a similar problem now, where someone could type `hil __init__` and get a strange result.

**Comment by naved001** [comment]
I have made some changes to use subparsers. My design still has some flaws. I am still unsure how to do multiple levels of subcommands (for example, see the `node` and `list` subcommands in the diff).

What I had done earlier, i.e. without the use of sub parsers was more intuitive for me tbh. Maybe I am just using sub-parsers the wrong way. I am gonna set this aside for now and come back to it in a while. 

Some questions I had:

* Should we have `hil show node <node-name>` and `hil node <node-name> show` at the same time?
* Should we support `hil node <node-name> connect project <project-name>` and `hil project <project-name> connect node <node-name>`

* In addition to list and show subcommands, should we have a `register` and `delete` subcommand too?
  - `hil register/delete node <node-name>`
  - `hil register/delete project <project-name>`
  - `hil register/delete nic <nic-name> <node-name>`
  - `hil register/delete network <args>`
  - `hil register/delete user <args>`

@xuhang57 feel free to poke around with argparse and/or my code. Here's the [diff](https://github.com/CCI-MOC/hil/compare/master...naved001:introduce-argparse?expand=1)

cc// @zenhack

**Comment by xuhang57** [comment]
In short, why we are not looking in to click? Here is why click based on its doc: http://click.pocoo.org/5/why/

I am still poking around and trying to figure out what would be the best approach. But it seems to me that `click` should be able to meet our requirements and it **seems** to work well with the subcommands

**Comment by zenhack** [comment]
Yeah, I've used click before, and it's nice; if we want to go that route I'm on-board. I do remember their subcommand support being pretty good.

**Comment by xuhang57** [comment]
And through reading this documentation: https://docs.openstack.org/python-openstackclient/latest/cli/commands.html

I am thinking of we should have something like:

```bash
hil node show <node>
hil project list
hil network show <network>
```

(Still reading and will expand the list)

**Comment by zenhack** [comment]
I'd rather not have alternatives like `node register` vs `register node`.

I like the idea of having the object first, so as @xuhang57 suggests, we'd do `hil node show <node>`. This mirrors the structure of the client library, which is nice.

**Comment by xuhang57** [comment]
Thinking of sitting down with @naved001 , if he is available tomorrow, and pair-programming on the `cli` with `click`. And send a PR so that Ian and other folks(spring break week so many people are not here) can review.

**Comment by zenhack** [comment]
Sounds good.

**Comment by naved001** [comment]
Thanks for the feedback @xuhang57 

The reason we rejected click was because it was based on optparse which is old, and apparently Argparse is the future. I, too, would like to spend time building the CLI rather than just playing with the python APIs for argparse. @zenhack is there any disadvantage of using a tool that's built on something which is deprecated (besides not getting the latest and greatest features)?

**Comment by zenhack** [comment]
Doesn't worry me. I don't remember this conversation.

Quoting Naved Ansari (2018-03-05 14:57:15)

**Comment by naved001** [comment]
> I don't remember this conversation.

Ah, I probably discussed this in person with @henn and others.

> hil node show console <node>
> hil node start console <node>
> hil node stop console <node>

I am thinking it should be `hil node console show/start/stop` because we are performing an action on node's console.

or, `hil node show-console <node>`

or `hil console show/start/stop <node-name>` because console only belongs to a node anyway.

all these are simpler to implement using Click. Will think more.

**Comment by naved001** [comment]
Also,

> hil switch list
> hil switch show <switch>
> hil switch show port <switch> <port>
> hil switch register <switch> <subtype> <args>
> hil switch delete <switch>
> hil switch connect nic <switch> <port> <node> <nic>
> hil switch detach nic <switch> <port>
> hil switch delete port <switch> <port>
> hil switch register port <switch> <port>
> hil switch revert port <switch> <port>

I think ports should have their own subcommand.

**Comment by xuhang57** [comment]
as we discussed, based on the [guideline/convention ](https://docs.openstack.org/python-openstackclient/latest/cli/commands.html)that I have been using, we have a problem that is we have a two-word `action` and usually, a convention (I guess) is to use a one-word action.

For example,

`show console` should be an action but it contains two words.
`connect network` should be an action and an object so it doesn't have a problem.

I don't know whether we should turn all two-word actions to one-word or we should use a `_` to connect them so it looks like a one-word. Any suggestion? 

Also, the guide I am using does not mean that is what we should do but it gives us an idea about how the community is doing. I haven't digged the official python documentation about how CLI should be constructed in general. Everyone is more than welcome to suggest a guide/convention/standard that we should follow. 

And I agree with @naved001 that we should turn some `switch` cli commands to that of `port`

**Comment by xuhang57** [comment]
here is an example how OS is dealing the situation: https://docs.openstack.org/mitaka/cli-reference/nova.html

`nova get-vnc-console <server> <console-type>`

They use dash to solve two-word action. Equivalently, we have

`hil node show-console <node>`

**Comment by zenhack** [comment]
Yeah, my inclination is to do similarly.

Quoting Lucas H. Xu (2018-03-06 13:57:02)

**Comment by naved001** [comment]
okay, one more thing that I wanted to run by you guys.

So the cli can't differentiate between `hil node register` and `hil node register nic` you can see the problem right away. I can either do `hil node register-nic` or `hil node nic register`. The advantage of the latter is I can then have subcommands for nics too, so I can do things like `hil node nic register/delete/show`, the downside is it reads like `hil object sub-object action`.

**Comment by zenhack** [comment]
I don't have a strong opinion.

Quoting Naved Ansari (2018-03-06 17:24:07)

> the downside is it reads like hil object sub-object action.

Not sure this is actually a downside; is that not accurate?

**Comment by naved001** [comment]
> Not sure this is actually a downside; is that not accurate?

I mean, the [openstack guide](https://docs.openstack.org/python-openstackclient/pike/cli/commands.html) was suggesting things like `object1 action object2 <args>` which I thought might be the general consensus, but it's not the only way. I am just inclined to make pool together all the objects first followed by the operations because I find it more consistent. I'll write what I think is the best and then y'all can look at it.

Also thanks for the prompt responses on this issue, makes it quicker to proceed.

**Comment by xuhang57** [comment]
I thought you were able to make `hil node connect network <node> <network>`?

I am fine with either way actually. As long as it makes sense to human ;)

**Comment by xuhang57** [comment]
we have `set-admin` so we can set a user to become an admin. Do we have something that we can set an admin to become a regular user? Do we need it?

@zenhack @naved001

**Comment by naved001** [comment]
> Do we have something that we can set an admin to become a regular user? Do we need it?

Nah, we purposely avoided that. Moreover, we want to be more careful when adding any extension specific API calls.

**Comment by xuhang57** [comment]
cool~, avoiding admins to attack each other ;)

Thanks!

**Comment by naved001** [comment]
Fixed by #969
